website/docs/tutorials/07-testing/03-fuzzing/README.md

```diff
@@ -0,0 +1,248 @@
+[PASS] tests/test_main.cairo test_withdraw (steps=129)
+       range_check_builtin=1
+
+11:28:43 [INFO] Test suites: 1 passed, 1 total
+11:28:43 [INFO] Tests:       1 passed, 1 total
+```
+
+### Generalizing the test
+
+This unit test does test that we can withdraw "some" amount from our safe.
+However, can we be sure that it works for all amounts, not just 1?
+
+The general property here is: given a safe balance, when we withdraw some amount from it,
+we should get reduced balance in the safe, and it should not be possible to withdraw more than we
+have.
+
+Protostar will run any test that takes parameters in fuzz testing mode, so let's rewrite our unit
+test.
+We need to take the `Cannot withdraw more than stored.` error, so we also add a call to
+the [`expect_revert`](../02-cheatcodes/expect-revert.md) cheatcode if needed.
+
+```cairo title="tests/test_main.cairo
+@external
+func test_withdraw{
+    syscall_ptr : felt*,
+    pedersen_ptr : HashBuiltin*,
+    range_check_ptr
+}(
+    amount : felt
+):
+    alloc_locals
+    let (pre_balance_ref) = balance.read()
+    local pre_balance = pre_balance_ref
+
+    %{
+        if ids.amount > ids.pre_balance:
+            expect_revert(error_message="Cannot withdraw more than stored.")
+    %}
+    withdraw(amount)
+
+    let (post_balance) = balance.read()
+    assert post_balance = pre_balance - amount
+
+    return ()
+end
+```
+
+If we run the test now, we can see that Protostar runs a fuzz test, but it fails for high values
+of `amount`:
+
+```text title="$ protostar test"
+[FAIL] tests/test_main.cairo test_withdraw
+[type] TRANSACTION_FAILED
+[code] 39
+[messages]:
+— Cannot withdraw more than stored.
+[details]:
+<REDACTED>/starkware/cairo/common/math.cairo:40:5: Error at pc=0:0:
+Got an exception while executing a hint.
+    %{
+    ^^
+Cairo traceback (most recent call last):
+tests/test_main.cairo:22:6: (pc=0:141)
+func test_withdraw{syscall_ptr : felt*, range_check_ptr, pedersen_ptr : HashBuiltin*}(
+     ^***********^
+tests/test_main.cairo:36:5: (pc=0:125)
+    withdraw(amount)
+    ^**************^
+Error message: Cannot withdraw more than stored.
+<REDACTED>/src/main.cairo:23:9: (pc=0:63)
+        assert_nn(new_res)
+        ^****************^
+
+Traceback (most recent call last):
+  File "<REDACTED>/starkware/cairo/common/math.cairo", line 43, in <module>
+    assert 0 <= ids.a % PRIME < range_check_builtin.bound, f'a = {ids.a} is out of range.'
+AssertionError: a = 3618502788666131213697322783095070105282824848410658236509717448704103819025 is out of range.
+[falsifying example]:
+amount = 340282366920938463463374607431768211456
+
+11:41:48 [INFO] Test suites: 1 failed, 1 total
+11:41:48 [INFO] Tests:       1 failed, 1 total
+11:41:48 [INFO] Seed:        4258368192
+```
+
+### Fixing the bug
+
+The test fails because if `amount` value is high enough, because it is a `felt`, when it is
+interpreted as a signed number in range comparisons, it overflows.
```

Comment:
This sentence is a bit weird. What do you think about:

```suggestion
The test fails because `amount` has `felt` type so its value can be negative. If smallest possible `felt` value is substracted from `balance` it causes `felt` overflow.
```